internal/datastore/mongo_store.go

```diff
@@ -114,14 +144,86 @@ func (mongodb MongoDB) DeleteQueue(id db.QueueId) error {
 	return nil
 }
 
-func (mongodb MongoDB) AddTokenToQueue(db.QueueId, db.Token) {
-	panic("Not implemented")
+func (mongodb MongoDB) AddTokenToQueue(id db.QueueId, token db.Token) (db.TokenId, error) {
+	token.Id = ""
+	token.QueueId = id
+
+	// Lock queue to ensure 2 concurrent add tokens don't happen
+	queueMutex.Lock()
+
+	// if there were 2 concurrent calls to this function, 2 tokens might get the same number
+	max, err := mongodb.GetMaxToken(id)
+
+	if err != nil {
+		log.Fatal(err)
+		queueMutex.Unlock()
+		return token.Id, err
+	}
+
+	token.TokenNumber = max + 1
+
+	result, err := mongodb.Token.InsertOne(context.TODO(), token)
+
+	queueMutex.Unlock()
+
+	if err != nil {
+		return token.Id, err
+	}
+
+	stringId := result.InsertedID.(primitive.ObjectID).Hex()
+
+	return db.TokenId(stringId), nil
 }
 
-func (mongodb MongoDB) ReadToken(db.TokenId) {
-	panic("Not implemented")
+func (mongodb MongoDB) GetMaxToken(id db.QueueId) (uint32, error) {
+	filter := bson.D{
+		// queueid is a string this time
+		{Key: "queueid", Value: id},
+	}
+
+	sort := bson.D{
+		{Key: "tokennumber", Value: -1},
+	}
+
+	findOptions := options.Find()
+	findOptions.SetSort(sort)
+	findOptions.SetLimit(1)
+
+	cursor, err := mongodb.Token.Find(context.TODO(), filter, findOptions)
+
+	if err != nil {
+		return 0, err
+	}
+
+	var tokens []db.Token
+
+	if err = cursor.All(context.TODO(), &tokens); err != nil {
+		log.Println(err)
+		return 0, err
+	}
+
+	if len(tokens) <= 0 {
+		return 0, nil
+	}
+
+	return tokens[0].TokenNumber, nil
+}
+
+func (mongodb MongoDB) ReadToken(id db.TokenId) (db.Token, error) {
+	tokenId, _ := primitive.ObjectIDFromHex(string(id))
+
+	var result db.Token
+
+	err := mongodb.Token.FindOne(context.TODO(), bson.M{"_id": tokenId}).Decode(&result)
+
+	if err != nil {
+		log.Fatal(err)
+		return result, err
```

Comment:
Did you have a reason to return `result` here? Wouldn't `return nil, err` be better?

Reply:
The return type is `db.Token` which cannot have a nil value. So we need to return something. In this case, it will be just an empty struct